rfcs/RFC-Upload-From-Dropbox.md

```diff
@@ -0,0 +1,358 @@
+
+While streaming the file, the compute resource will periodically check if the upload has been canceled using the upload
+table. If the upload status is cancel pending, the compute resource will stop the stream, delete any data that has been uploaded,
+mark the job as complete in the queue, and set the status of the upload to cancelled.
+
+#### 7. Retry
+
+If the upload fails for any reason, the compute resource will mark the status of the upload as waiting, and return the upload
+job to the queue.
+
+#### 8. Failed Uploads
+
+If an upload fails enough times, it will be removed from the download queue. The upload status will change to failed.
+
+### Components
+
+This section describes the purpose of components that make up the upload design.
+
+#### Data Portal Browser App
+
+The DP Browser App is the frontend of the system. It is the primary method users have to interact with the Data Portal.
+It will be used to authenticate the user and interact with the DP Backend to upload files.
+
+#### Data Portal Backend
+
+The DP Backend is responsible for handling API requests. It will enqueue shared links for downloading, and update the
+upload status. This additional functionality can be built into the existing DP Backend Lambda.
+
+The backend will also be responsible for performing a simple validation of the file being uploaded. The validation is
+performed during the initial request to upload. The validation consists of verifying the extension of the file is
+".h5ad".
+
+#### Download Queue
+
+The Download Queue is an [AWS SQS](https://aws.amazon.com/sqs/) that will track pending download jobs.
+The download jobs are used by the upload service to know where to download the file from before uploading them to S3.
+
+##### Download Job Entry
+
+These are the fields that will be in the download job placed in the queue:
+
+- Link - the link to download. Use to retrieve the file from CSP.
+- submission_id - identifies the submission. Used to determine the storage location.
+- dataset_id - identifies the dataset. Used to determine the storage location
+- file_name - the name of the file being downloaded
+
+#### Upload Service
+
+A computing resource within the AWS cloud. It will receive download jobs from the Download Queue. Those jobs contain
+the information need by the compute resource to download a file from DropBox and upload it to S3. The computer resource
+will also verify the integrity of the download and upload. The compute resource will also change the state of the
+upload in the Upload table as needed.
+
+While the download job is processing, the computing resource with occasionally poll the upload table for a cancel
+pending status. If the cancel pending status is set, the computing resource will delete any data uploaded to S3,
+remove the download job from the queue, and change the status of the upload to canceled.
+
+If an error occurs while processing the upload, the compute resource will return the download job to the queue and
```

Comment:
This should depend on the type of error.    There are transient errors which cause a failure, but may work if you retry (e.g. network errors).   There are unrecoverable errors (bad link, access denied, file too large, or not the write suffix, etc), where retrying will not help.   For the unrecoverable errors, we should provide an error message back to the user, and not retry.

Reply:
Added a message field that will be returned if the status is failed.